File: kinv/core/parse/footprint.py
"""What a footprint reference like ``Capacitor_SMD:C_0603_1608Metric`` says about the part."""

from __future__ import annotations

import re

# Library nickname → class. Checked before the name, and longest first.
_LIBRARY_CLASSES: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"^Capacitor", re.I), "capacitor"),
    (re.compile(r"^Resistor", re.I), "resistor"),
    (re.compile(r"^Inductor", re.I), "inductor"),
    (re.compile(r"^Ferrite", re.I), "ferrite"),
    (re.compile(r"^(Diode|LED)", re.I), "diode"),
    (re.compile(r"^(Transistor|Package_TO)", re.I), "transistor"),
    (re.compile(r"^Connector", re.I), "connector"),
    (re.compile(r"^(Button_Switch|Switch)", re.I), "switch"),
    (re.compile(r"^(Crystal|Oscillator)", re.I), "crystal"),
    (re.compile(r"^Transformer", re.I), "transformer"),
    (re.compile(r"^Fuse", re.I), "fuse"),
    (re.compile(r"^Relay", re.I), "relay"),
    (re.compile(r"^TestPoint", re.I), "testpoint"),
    (re.compile(r"^MountingHole|^Mounting_", re.I), "mounting"),
    (re.compile(r"^Package_", re.I), "ic"),
]

# Footprint-name prefix → class, for libraries that say nothing useful.
_NAME_CLASSES: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"^C_"), "capacitor"),
    (re.compile(r"^CP_"), "capacitor"),
    (re.compile(r"^R_"), "resistor"),
    (re.compile(r"^L_"), "inductor"),
    (re.compile(r"^FB_"), "ferrite"),
    (re.compile(r"^(D_|LED_)"), "diode"),
    (re.compile(r"^Q_"), "transistor"),
    (re.compile(r"^TestPoint"), "testpoint"),
    (re.compile(r"^MountingHole"), "mounting"),
    (re.compile(r"^PinSocket|^PinHeader|^USB_|^Conn_"), "connector"),
]

_CHIP = re.compile(r"_([0-9]{4})_([0-9]{4})Metric")
_VARIANT = re.compile(r"_(HandSolder|Pad[0-9.x]+mm(?:_HandSolder)?|Castellated|ReverseGeometry)", re.I)


def parse_footprint(raw: str) -> dict | None:
    """``{raw, library, name, cls, chip, variant}``, or None for an empty cell."""
    trimmed = raw.strip()
    if trimmed == "":
        return None

    colon = trimmed.find(":")
    library = trimmed[:colon] if colon >= 0 else None
    name = trimmed[colon + 1 :] if colon >= 0 else trimmed

    cls = "unknown"
    if library is not None:
        for pattern, candidate in _LIBRARY_CLASSES:
            if pattern.search(library):
                cls = candidate
                break
    if cls == "unknown":
        for pattern, candidate in _NAME_CLASSES:
            if pattern.search(name):
                cls = candidate
                break

    chip_match = _CHIP.search(name)
    chip = {"imperial": chip_match.group(1), "metric": chip_match.group(2)} if chip_match else None

    variant_match = _VARIANT.search(name)
    variant = variant_match.group(1) if variant_match else None

    return {"raw": trimmed, "library": library, "name": name, "cls": cls, "chip": chip, "variant": variant}


def package_of(fp: dict | None) -> str | None:
    """The package token in canonical keys.

    The imperial chip code where there is one (``0603``), otherwise the
    footprint name, which for a specific part is the most honest identity
    available.
    """
    if not fp:
        return None
    return fp["chip"]["imperial"] if fp["chip"] else fp["name"]

File: src/core/parse/footprint.ts
import type { ComponentClass } from "./refdes.js";

export type ParsedFootprint = {
  raw: string;
  /** Library nickname, `Capacitor_SMD` in `Capacitor_SMD:C_0603_1608Metric`. */
  library: string | undefined;
  /** Footprint name within the library. */
  name: string;
  /** Class the footprint implies. Compared against the refdes to catch mix-ups. */
  cls: ComponentClass;
  /** Two-terminal chip size, when the name encodes one. */
  chip: { imperial: string; metric: string } | undefined;
  /** `HandSolder`, `Pad1.33x2.70mm` etc. — same part, different land pattern. */
  variant: string | undefined;
};

/** Library nickname → class. Checked before the name, and longest first. */
const libraryClasses: ReadonlyArray<readonly [RegExp, ComponentClass]> = [
  [/^Capacitor/i, "capacitor"],
  [/^Resistor/i, "resistor"],
  [/^Inductor/i, "inductor"],
  [/^Ferrite/i, "ferrite"],
  [/^(Diode|LED)/i, "diode"],
  [/^(Transistor|Package_TO)/i, "transistor"],
  [/^Connector/i, "connector"],
  [/^(Button_Switch|Switch)/i, "switch"],
  [/^(Crystal|Oscillator)/i, "crystal"],
  [/^Transformer/i, "transformer"],
  [/^Fuse/i, "fuse"],
  [/^Relay/i, "relay"],
  [/^TestPoint/i, "testpoint"],
  [/^MountingHole|^Mounting_/i, "mounting"],
  [/^Package_/i, "ic"],
];

/** Footprint-name prefix → class, for libraries that say nothing useful. */
const nameClasses: ReadonlyArray<readonly [RegExp, ComponentClass]> = [
  [/^C_/, "capacitor"],
  [/^CP_/, "capacitor"],
  [/^R_/, "resistor"],
  [/^L_/, "inductor"],
  [/^FB_/, "ferrite"],
  [/^(D_|LED_)/, "diode"],
  [/^Q_/, "transistor"],
  [/^TestPoint/, "testpoint"],
  [/^MountingHole/, "mounting"],
  [/^PinSocket|^PinHeader|^USB_|^Conn_/, "connector"],
];

const chipPattern = /_(\d{4})_(\d{4})Metric/;
const variantPattern = /_(HandSolder|Pad[\d.x]+mm(?:_HandSolder)?|Castellated|ReverseGeometry)/i;

export function parseFootprint(raw: string): ParsedFootprint | undefined {
  const trimmed = raw.trim();
  if (trimmed === "") return undefined;

  const colon = trimmed.indexOf(":");
  const library = colon >= 0 ? trimmed.slice(0, colon) : undefined;
  const name = colon >= 0 ? trimmed.slice(colon + 1) : trimmed;

  let cls: ComponentClass = "unknown";
  if (library !== undefined) {
    for (const [re, c] of libraryClasses) {
      if (re.test(library)) {
        cls = c;
        break;
      }
    }
  }
  if (cls === "unknown") {
    for (const [re, c] of nameClasses) {
      if (re.test(name)) {
        cls = c;
        break;
      }
    }
  }

  const chipMatch = chipPattern.exec(name);
  const chip = chipMatch
    ? {
        imperial: chipMatch[1] as string,
        metric: chipMatch[2] as string,
      }
    : undefined;

  const variantMatch = variantPattern.exec(name);
  const variant = variantMatch ? (variantMatch[1] as string) : undefined;

  return { raw: trimmed, library, name, cls, chip, variant };
}

/**
 * The package token used in canonical keys: the imperial chip code where there
 * is one (`0603`), otherwise the footprint name, which for a specific part is
 * the most honest identity available.
 */
export function packageOf(fp: ParsedFootprint | undefined): string | undefined {
  if (!fp) return undefined;
  return fp.chip?.imperial ?? fp.name;
}
